src/mlnet/Telemetry/Telemetry.cs

```diff
@@ -1,144 +1,239 @@
+            {
+                sb.Append(transformNode.Name);
+                sb.Append(",");
+            }
+            if (trainerNode != null)
+            {
+                sb.Append(trainerNode.Name);
+                sb.Append("{");
+                var serializedHyperparams = trainerNode.Properties
+                    .Where(p => SweepableParams.AllHyperparameterNames.Contains(p.Key))
+                    .Select(p => $"{p.Key}: {p.Value}");
+                sb.Append(string.Join(", ", serializedHyperparams));
+                sb.Append("}");
+            }
+            return sb.ToString();
         }
 
-        private void InitializeTelemetry()
+        private static void InitializeTask()
         {
             try
             {
+                // Since the user didn't yet have a chance to read the above message and decide to opt out,
+                // don't log any telemetry on the first invocation.
+                if (CheckFirstTimeUse())
+                {
+                    _enabled = false;
+                    return;
+                }
+
                 _client = new TelemetryClient();
                 _client.InstrumentationKey = InstrumentationKey;
                 _client.Context.Device.OperatingSystem = RuntimeEnvironment.OperatingSystem;
 
-                // we don't want hostname etc to be sent in plain text.  
-                // these need to be set to some non-empty values to override default behavior.
+                // We don't want hostname etc to be sent in plaintext.  
+                // These need to be set to some non-empty values to override default behavior.
                 _client.Context.Cloud.RoleInstance = "-"; 
                 _client.Context.Cloud.RoleName = "-";
 
                 _commonProperties = new TelemetryCommonProperties().GetTelemetryCommonProperties();
+                // Add a session ID to each log sent during the life of this process
+                _commonProperties["SessionId"] = Guid.NewGuid().ToString();
             }
             catch (Exception e)
             {
                 _client = null;
-                // we dont want to fail the tool if telemetry fails.
+                // We dont want to fail the tool if telemetry fails.
                 Debug.Fail(e.ToString());
             }
         }
 
-        private void TrackEventTask(
+        private static void TrackEventTask(
             string eventName,
             IDictionary<string, string> properties,
-            IDictionary<string, double> measurements)
+            TimeSpan? duration,
+            Exception ex)
         {
-            if (_client == null)
+            if (_client != null)
             {
-                return;
+                try
+                {
+                    var eventProperties = GetEventProperties(properties, duration, ex);
+                    _client.TrackEvent(eventName, eventProperties);
+                    _client.Flush();
```

Comment:
Thinking out loud here.  I am not convinced doing a Flush() here makes sense anymore.  Some events like iteration events with a high probability of the process continuing to run would do well with Flush() right there.  But in worst case scenario where we posted some events and now want to exit, doing a Flush() on each of them would likely make it not fit in 200ms time box that you have.  I think a more generic way would be to have an event queue that would accept events and only do a flush if explicitly told so (on exit of the process) or if there's been say more than 100ms idle time with the queue empty.

Reply:
worth a look at how .NET CLI handles this now or how it did handle it before they implemented persisting unsent telemetry between invocations

Reply:
"But in worst case scenario where we posted some events and now want to exit, doing a Flush() on each of them would likely make it not fit in 200ms time box that you have."
I could definitely be wrong, but I was thinking the opposite is true. These tasks are all done in parallel. The earlier we start each network request, the more likely it is that telemetry event has been sent in the case of an abrupt exit